### ISSUE TITLE
Vertical alignment

### ISSUE THREAD

**Issue by rleonid** [hammerlab/style-guides#2]
I've become a big fan of vertically aligning data-structure definitions and let bindings. So I strong prefer:

``` OCaml
type exploration_state = {
  build_process_details : bool;
  target_filter         : (Target.t -> bool) * Log.t;
  current_target        : Target.id option;
  condition_details     : bool;
}

let create_state () = { 
  build_process_details = false;
  condition_details     = false;
  target_filter         = (fun _ -> true), Log.(s "No-filter, see them all");
  current_target        = None
}
```

to 

``` OCaml
type exploration_state = {
  build_process_details: bool;
  target_filter: (Target.t -> bool) * Log.t;
  current_target: Target.id option;
  condition_details: bool;
}

let create_state () =
  {build_process_details = false;
    condition_details = false;
    target_filter = (fun _ -> true), Log.(s "No-filter, see them all");
    current_target = None}
```

similarly, I like

``` OCaml
let foo a b c =
    let x   = a () in
    let xx  = b x in
    let y   = c xx in
    ...
```

as opposed to

``` OCaml
let foo a b c =
    let x = a () in
    let xx = b x in
    let y = c xx in
```

I find it easier to read code as performing a function this way. I know that I resisted this style for quiet some time in my own code, but now find it very helpful. Let me know what you guys think!

**Comment by hammer** [comment]
Can you enforce this style easily with `ocp-indent` or your own emacs plugin?

**Comment by smondet** [comment]
`ocp-indent`'s specification so far is to touch only the whitespace in front of a line.

Emacs and Vim have align with regexp functions/plugins, but nothing is automatic

Some people complain that aligned code is more difficult to maintain (you have to keep alignment in sync which can be painful, and then your git-diffs are bigger than they should).

I like it in some cases and don't care in most others. I don't think it should be enforced in the style-guide.

**Comment by rleonid** [comment]
@hammer , yes, @smondet is spot on; it doesn't seem like `ocp-indent` has these options. All of it's config options are for how far to indent. It does have `align_ops` and `align_params` but those are hooks to override the default indentation so that lines align based upon the previous line's `=` or `->`.

I agree with Seb that it can be more difficult to maintain, so something like

``` OCaml
type exploration_state = {
  build_process_details : bool;
  target_filter         : (Target.t -> bool) * Log.t;
  current_target        : Target.id option;
  condition_details     : bool;
}
```

can go 'full haskell' and be written

``` OCaml
type exploration_state = { build_process_details : bool
                         ; target_filter         : (Target.t -> bool) * Log.t
                         ; current_target        : Target.id option
                         ; condition_details     : bool;
                         }
```

and that's fairly easy to maintain since the lines are much more independent of each other.

I agree that it shouldn't be an enforced style. I just want to get feedback on that style.

**Comment by smondet** [comment]
I don't understand why the “haskell-like” version would be easier to maintain, you still have to update other lines to keep alignment, right?
(it is also very non-idiomatic, and those `;` look typographically awful in front of the line, don't they? :))

**Comment by smondet** [comment]
Here is an example where the alignment makes the type so far from the name that it is difficult to do the matching
https://github.com/janestreet/async_unix/blob/master/lib/writer0.ml#L38

(or maybe you need horizontal lines guiding your eyes on your text editor...)

**Comment by rleonid** [comment]
Sure, that's a bit of an extreme example. But honestly, I don't find it particularly difficult to read, modulo the complexity of the structure they're creating. I aslo don't think that they type for 'id' is particularly relevant. But consider the trade-of on this line: https://github.com/janestreet/async_unix/blob/master/lib/writer0.ml#L72 , where they group the buffer and the indexer's into it. I find that particularly helpful. Yes, the grouping is actually performed by the comments and separating white lines, but rearranging those definitions is easier once you move the semi-colons to the front.

**Comment by smondet** [comment]
Actually the semicolons in the front make the "doc-attachment" very confusing, the comment on `buf` seems to be for the previous field since it is before the separator (that's also what a doc-parser would understand, if only those comments where `(**`).

That huge whitespace gap makes my eyes go back to the beginning of the line to slowly and carefully try to follow an imaginary horizontal line to reach the type definition.
